specs/agents/tracing-instrumentation-messaging.md

```diff
@@ -0,0 +1,103 @@
+## Messaging Systems
+
+The instrumentation of messaging systems includes:
+- Transaction creation on message reception, either as a child of the message-sending span, if the sending action is traced (meaning 
+distributed tracing support), or as root
+- Span creation for message sending action
+- Span creation on message *polling that occurs within a traced transaction*
+
+### Passive vs. active message reception
+
+Message send/publish events can only be captured as spans if occurring within a traced transaction.
+Message consumption can be divided into two types: passive, where you would implement a listener that is called once a message is available, 
+and active, where the queue/topic is being polled (blocking or non-blocking). 
+
+Passive consumption typically results in a `messaging` transaction and is pretty straightforward to trace - start at entry and end at exit. 
+
+Message polling can be done within a traced transaction, in which case it should result in a messaging span, or it can be the initiating 
+event for a message handling flow. Capturing polling spans is also mostly straightforward. 
+For polling-based transactions, our goal is to capture the message handling flow, which typically *starts after the polling action exits*, 
+returning a message. This may be tricky if the handling flow is not implemented within a well defined API. Use other agent implementation
+as reference.
+The agent should not create a transaction based on polling APIs if the polling action did not result with a message (as opposed to 
+polling spans, where we want to capture such as well).
+
+### Typing
+
+- Transactions: 
+  - `transaction.type`: `messaging`
+- Spans: 
+  - `span.type`: `messaging` 
+  - `span.subtype`: the name of the framework - e.g. `jms`, `kafka`, `rabbitmq` 
+  - `span.action`: `send`, `receive` or `poll`
+  
+### Naming
+
+Transaction and span names may* follow this pattern: `<MSG-FRAMEWORK> SEND/RECEIVE/POLL to/from <QUEUE-NAME>`.
+Examples:
+- `JMS SEND to MyQueue`
+- `RabbitMQ RECEIVE from MyQueue`**
+- `RabbitMQ POLL from MyExchange`**
+
+Agents may deviate from this pattern, as long as they ensure a proper cardinality is maintained, that is- neither too low nor too high. 
+For example, agents may choose to name all transactions/spans reading-from/sending-to temporary queues equally. 
+On the other end, agents may choose to append a cardinality-increasing factor to the name, like the routing key in RabbitMQ.
+
+\* Java agent's instrumentation for Kafka does not follow this pattern at the moment.
+
+#### \** RabbitMQ naming specifics 
+
+In RabbitMQ, queues are only relevant in the receiver side, so the exchange name is used instead for sender spans.
+When the default exchange is used (denoted with an empty string), it should be replaced with `<default>`. 
+
+Agents may add an opt-in config to append the routing key to the name as well, for example: `RabbitMQ RECEIVE from MyExchange\58D7EA987`.
```

Comment:
[minor] maybe use a `/` instead of a `\` I don't know why but it feels better and less window-ish or escape-character-in-a-shell-script-ish.